engine/servermaster/cluster/election.go

```diff
@@ -131,14 +133,18 @@ func (e *EtcdElection) doCampaign(ctx context.Context, selfID NodeID, timeout ti
 		return nil, nil, derror.WrapError(derror.ErrMasterEtcdElectionCampaignFail, err)
 	}
 	retCtx := newLeaderCtx(ctx, e.session)
+
+	var resignOnce sync.Once
 	resignFn := func() {
-		resignCtx, cancel := context.WithCancel(ctx)
-		defer cancel()
-		defer retCtx.OnResigned()
-		err := e.election.Resign(resignCtx)
-		if err != nil {
-			log.Warn("resign leader failed", zap.Error(err))
-		}
+		resignOnce.Do(func() {
```

Comment:
In the previous implementation of `leaderLoop`, it will. Now, although it won't be called more than once, it's better to make it idempotent like the original etcd api.